flow/flow.go

```diff
@@ -0,0 +1,303 @@
+package flow
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/micro/go-micro/v2/broker"
+	"github.com/micro/go-micro/v2/client"
+	"github.com/micro/go-micro/v2/logger"
+)
+
+type Flow interface {
+	// Init flow with options
+	Init(...Option) error
+	// Get flow options
+	Options() Options
+	// Create step in specific flow
+	CreateStep(flow string, step *Step) error
+	// Delete step from specific flow
+	DeleteStep(flow string, step *Step) error
+	// Replace step in specific flow
+	ReplaceStep(flow string, oldstep *Step, newstep *Step) error
+	// Lookup specific flow
+	Lookup(flow string) ([]*Step, error)
+	// Execute specific flow and returns request id and error, optionally fills rsp
+	Execute(flow string, req interface{}, rsp interface{}, opts ...ExecuteOption) (string, error)
+	// Resume specific paused flow execution by request id
+	Resume(flow string, reqID string) error
+	// Pause specific flow execution by request id
+	Pause(flow string, reqID string) error
+	// Abort specific flow execution by request id
+	Abort(flow string, reqID string) error
+	// State show state specific flow execution by request id
+	State(flow string, reqID string) (string, error)
+	// Result get result of the flow step
+	Result(flow string, reqID string, step *Step) ([]byte, error)
+	// Stop executor and drain active workers
+	Stop() error
+}
+
+type Step struct {
+	// name of step
+	ID string
+	// Retry count for step
+	Retry int
+	// Timeout for step
+	Timeout int
+	// Step operation to execute
+	Operation Operation
+	// Which step use as input
+	Input string
+	// Where to place output
+	Output string
+	// Steps that are required for this step
+	After []string
+	// Steps for which this step required
+	Before []string
+	// Step operation to execute in case of error
+	Fallback Operation
+}
+
+func (s *Step) Name() string {
+	return s.ID
+}
+
+func (s *Step) Id() string {
+	return s.ID
+}
+
+func (s *Step) String() string {
+	return s.ID
+	//return fmt.Sprintf("step %s, ops: %s, requires: %v, required: %v", s.ID, s.Operations, s.Requires, s.Required)
+}
+
+type Option func(*Options)
+
+type Options struct {
+	// Number of worker goroutines
+	Concurrency int
+	// Preallocate worker goroutines
+	Prealloc bool
+	// If no workers available no blocking
+	Nonblock bool
+	// Wait completiong before stop
+	Wait bool
+	// StateStore is used for flow state marking
+	StateStore DataStore
```

Comment:
State/Data/Flow stores seem redundant. We should only have to pass in one store.

Reply:
no, i'm not add docs but this is must be different stores. 
Data - is []byte blob, and can be large, so it useful to store sometimes in s3 capable storage.
State - is []byte small blob, must be consistent and in-sync so its preferable distributed db or etcd or something.
Flow - is variable sized []byte blob that can be stored in db or in k/v or something. Mostly fast read store with ability to expose Timestamp for effective caching and transaction update support.

Reply:
Doesn't feel like the right way to impement it. Rather these are concerns taken care of by the implementation of the flow. The default flow being "in memory" perhaps and so its a non-concern. Then you might have an implementation that uses "aws" and so on.

Reply:
i think mostly for implementation not like change store, but about all logic inside - like pool of workers, parsing steps definition from store and so. For production mostly in memory not suitable - as you can be interrupted in the middle of execution, and you need continue/rollback laer (or via HA flow executor)